File: ruby/day03.rb
# # frozen_string_literal: true

open('../input/03.txt') do |f|
  prev = row = col = nil
  state = f.reduce(0) do |count, curr|
    curr = curr.chomp
    width = curr.size
    if prev.nil?
      row = col = 0
    else
      row += 1
      col += 3
      count += 1 if curr[col % width] == '#'
    end
    prev = curr
    count
  end
  p state
end

map = open('../input/03.txt') { |f| f.each_with_object([]) { |i, a| a << i.chomp } }

def count(map, right, down)
  height = map.size
  width = map.first.size

  count = 0

  col = 0
  (0...height).step(down) do |row|
    count += 1 if map[row][col % width] == '#'
    col += right
  end

  count
end

a = count(map, 1, 1)
b = count(map, 3, 1)
c = count(map, 5, 1)
d = count(map, 7, 1)
e = count(map, 1, 2)

p "#{a} * #{b} * #{c} * #{d} * #{e} = #{a * b * c * d * e}"

File: rust/src/bin/day03.rs
extern crate aoc2020;

use aoc2020::common::read_lines;

fn count_trees(map: &Vec<String>, right: usize, down: usize) -> usize {
    let height = map.len();
    let width = map.first().unwrap().len();

    let mut row = 0;
    let mut col = 0;

    let mut count = 0;
    while row < height {
        if map[row].chars().nth(col % width).unwrap() == '#' {
            count += 1;
        }
        row += down;
        col += right;
    }
    count
}

fn main() {
    if let Ok(lines) = read_lines("../input/03.txt") {
        let mut map: Vec<String> = Vec::new();
        for line in lines {
            if let Ok(ip) = line {
                map.push(ip);
            }
        }

        // Part 1
        println!("Right {}, down {}: {}", 3, 1, count_trees(&map, 3, 1));

        // Part 2
        println!(
            "Part 2: {}",
            count_trees(&map, 1, 1)
                * count_trees(&map, 3, 1)
                * count_trees(&map, 5, 1)
                * count_trees(&map, 7, 1)
                * count_trees(&map, 1, 2)
        );
    }
}
